operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -307,6 +314,85 @@ func (r *TenantReconciler) checkValidWorkspaces(ctx context.Context, tn *crownla
+func (r *TenantReconciler) deleteClusterNamespace(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string) error {
+	ns := v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}
+
+	err := utils.EnforceObjectAbsence(ctx, r.Client, &ns, "personal namespace")
+
+	if err != nil {
+		klog.Errorf("Error when deleting namespace of tenant %s -> %s", tn.Name, err)
+	}
+
+	return err
+}
+
+// checkNamespaceKeepAlive checks to see if the namespace should be deleted.
+func (r *TenantReconciler) checkNamespaceKeepAlive(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string, tenantNSKeepAlive time.Duration) (keepNsOpen bool, err error) {
+	// We check to see if last login was more than tenantNSKeepAlive in the past:
+	// if so, temporarily delete the namespace. We assume that a lastLogin of 0 occurs when a user is first created
+
+	// Calculate time elapsed since lastLogin (now minus lastLogin in seconds)
+	sPassed := time.Since(tn.Spec.LastLogin.Time)
+
+	klog.Infof("Last login of tenant %s was %s ago", tn.Name, sPassed)
+
+	// Attempt to get instances in current namespace
+	list := &crownlabsv1alpha2.InstanceList{}
+
+	if err := r.List(context.Background(), list, client.InNamespace(nsName)); err != nil {
+		return true, err
+	}
+
+	if sPassed > tenantNSKeepAlive { // seconds
+		klog.Infof("Over %s elapsed since last login of tenant %s: attempting to delete tenant namespace if not already deleted", tenantNSKeepAlive, tn.Name)
+		if len(list.Items) == 0 {
+			klog.Infof("No instances found in %s: namespace can be deleted", nsName)
+			return false, nil
+		}
+		klog.Infof("Instances found in namespace %s. Namespace will not be deleted", nsName)
+	} else {
+		klog.Infof("Under %s (limit) elapsed since last login of tenant %s: namespace is left as-is", tenantNSKeepAlive, tn.Name)
+	}
+
+	return true, nil
+}
+
+// Deletes namespace or updates the cluster resources.
+func (r *TenantReconciler) enforceClusterResources(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string, keepNsOpen bool) (nsOk bool, err error) {
+	nsOk = false // nsOk must be initialized for later use
+
+	if keepNsOpen {
+		nsOk, err = r.createOrUpdateClusterResources(ctx, tn, nsName)
+		if nsOk {
+			klog.Infof("Namespace %s for tenant %s updated", nsName, tn.Name)
+			tn.Status.PersonalNamespace.Created = true
+			tn.Status.PersonalNamespace.Name = nsName
+			if err != nil {
+				klog.Errorf("Unable to update cluster resource of tenant %s -> %s", tn.Name, err)
+				tnOpinternalErrors.WithLabelValues("tenant", "cluster-resources").Inc()
+			}
+			klog.Infof("Cluster resources for tenant %s updated", tn.Name)
+		} else {
+			klog.Errorf("Unable to update namespace of tenant %s -> %s", tn.Name, err)
+			tn.Status.PersonalNamespace.Created = false
+			tn.Status.PersonalNamespace.Name = ""
+			tnOpinternalErrors.WithLabelValues("tenant", "cluster-resources").Inc()
+		}
+	} else {
+		nsErr := r.deleteClusterNamespace(ctx, tn, nsName)
```

Comment:
```suggestion
		err = r.deleteClusterNamespace(ctx, tn, nsName)
```

Reply:
Resolved in [242da29](https://github.com/netgroup-polito/CrownLabs/pull/821/commits/242da290c92a7831c5801f500a7a3b352126f42d)